File: testcases/rust/minunit_test.rs
// In Cargo.toml, add:
// [dependencies]
// lazy_static = "1.4"

#[cfg(test)]
mod tests {
    use std::sync::Mutex;

    // We'll simulate a global variable that our setup() and teardown() modify.
    lazy_static::lazy_static! {
        static ref GLOBAL_VALUE: Mutex<i32> = Mutex::new(0);
    }

    /// Setup function: set the global value to 42.
    fn setup() {
        let mut val = GLOBAL_VALUE.lock().unwrap();
        *val = 42;
    }

    /// Teardown function: reset the global value to 0.
    fn teardown() {
        let mut val = GLOBAL_VALUE.lock().unwrap();
        *val = 0;
    }

    #[test]
    fn test_addition() {
        // Equivalent to: mu_assert_int_eq(4, 2 + 2);
        assert_eq!(2 + 2, 4);
    }

    #[test]
    fn test_double_equality() {
        // Equivalent to: mu_assert_double_eq(expected, result);
        let a = 3.141592653589793;
        let b = 3.141592653589792; // Very close value.
        let epsilon = 1e-12;
        assert!((a - b).abs() < epsilon, "Expected {} to be approximately equal to {}", a, b);
    }

    #[test]
    fn test_string_equality() {
        // Equivalent to: mu_assert_string_eq("minunit", "minunit");
        assert_eq!("minunit", "minunit");
    }

    #[test]
    fn test_boolean_check() {
        // Equivalent to: mu_check((5 - 3) == 2);
        assert!((5 - 3) == 2);
    }

    #[test]
    fn test_setup_teardown() {
        // Simulate running a test with setup and teardown.
        setup();
        {
            let val = GLOBAL_VALUE.lock().unwrap();
            assert_eq!(*val, 42, "Setup did not set GLOBAL_VALUE to 42");
        }
        teardown();
        {
            let val = GLOBAL_VALUE.lock().unwrap();
            assert_eq!(*val, 0, "Teardown did not reset GLOBAL_VALUE to 0");
        }
    }
}

File: testcases/c/minunit_test.c
/*
   test_minunit.c

   Example tests using minunit_minunit.h (minunit) to verify that the framework’s
   assertion macros and suite runner work as expected.
*/

#include <stdio.h>
#include <string.h>
#include "minunit_minunit.h"  /* Ensure this header is in your include path */

/* --- Global variable for testing setup/teardown --- */
static int global_value = 0;

/* --- Setup and Teardown functions --- */
static void test_setup(void) {
    /* For example, initialize global_value */
    global_value = 42;
}

static void test_teardown(void) {
    /* Reset global_value */
    global_value = 0;
}

/* --- Tests --- */

/* Test integer addition using mu_assert_int_eq */
MU_TEST(test_addition) {
    mu_assert_int_eq(4, 2 + 2);
}

/* Test double equality using mu_assert_double_eq */
MU_TEST(test_double_equality) {
    double a = 3.141592653589793;
    double b = 3.141592653589792;  /* very close */
    mu_assert_double_eq(a, b);
}

/* Test string equality using mu_assert_string_eq */
MU_TEST(test_string_equality) {
    mu_assert_string_eq("minunit", "minunit");
}

/* Test simple boolean check using mu_check */
MU_TEST(test_boolean_check) {
    mu_check( (5 - 3) == 2 );
}

/* Test that setup properly initializes a global variable */
MU_TEST(test_setup_teardown) {
    /* Since setup was run, global_value should be 42 */
    mu_assert_int_eq(42, global_value);
}

/* --- Test Suite --- */
static void all_tests(void) {
    /* Configure suite to use setup/teardown functions */
    MU_SUITE_CONFIGURE(test_setup, test_teardown);

    /* Run tests */
    MU_RUN_TEST(test_setup_teardown);  /* uses setup: global_value == 42 */
    MU_RUN_TEST(test_addition);
    MU_RUN_TEST(test_double_equality);
    MU_RUN_TEST(test_string_equality);
    MU_RUN_TEST(test_boolean_check);
}

/* --- Main --- */
int main(void) {
    MU_RUN_SUITE(all_tests);
    MU_REPORT();
    return MU_EXIT_CODE;
}
